### PR TITLE
[python-package] Allow to pass early stopping min delta in params

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

Currently, setting `min_delta` in the early stopping callback requires specifying the early stopping callback manually. It would be nice to continue passing all early-stopping-related parameters in `params`.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]

**Comment by borchero** [comment]
Finally ready now @jameslamb :smile:

**Review by jameslamb — CHANGES_REQUESTED** [review]
Looking into this, I just realized... `early_stopping_min_delta` is specific to the Python interface! I'm sorry for not noticing that earlier.

It isn't listed at https://lightgbm.readthedocs.io/en/latest/Parameters.html and isn't a valid property of the `LightGBM::Config` object on the C++ side (https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/config.h).

I guess because when that was added, it was only added on the Python side and not in this C++ code:

https://github.com/microsoft/LightGBM/blob/6a1ec44437bbc830e5c352aaea395994b28a79f5/src/boosting/gbdt.cpp#L456-L473

Given that, if it's going to show up in the `params` dictionary (which I get is valuable because it's the type of parameter you'd like to search a space of values for during hyperparam tuning), then I I think it should be popped off very early in `train()` / `cv()`, like this:

```python
early_stopping_min_delta = params.pop("early_stopping_min_delta", 0.0)
```

I haven't pulled your branch and tested, but strongly suspect LightGBM is raising this warning about an unrecognized parameter:

https://github.com/microsoft/LightGBM/blob/6a1ec44437bbc830e5c352aaea395994b28a79f5/src/io/config.cpp#L28

ALTERNATIVELY... this PR could be expanded by adding `early_stopping_min_delta` to the `Config` object.

That'd involve:

* adding an entry in https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/config.h
* running `python ./helpers/parameter_generator.py`
* modifying the tests from #6077 to be sure that parameter is making it through to model files and back:
  - https://github.com/microsoft/LightGBM/blob/6a1ec44437bbc830e5c352aaea395994b28a79f5/tests/python_package_test/test_engine.py#L1454
  - https://github.com/microsoft/LightGBM/blob/6a1ec44437bbc830e5c352aaea395994b28a79f5/R-package/tests/testthat/test_lgb.Booster.R#L828

----

@borchero @jmoralez what do you think?

* Option 1: add `early_stopping_min_delta` as a LightGBM-wide parameter (even if only the Python package currently uses it)
* Option 2: `.pop()` it off `params` in `train()` / `cv()` in the Python package

**Comment by jmoralez** [comment]
We didn't mark #2526 as done in #2302 because the `min_delta` should be implemented on the cpp side ([ref](https://github.com/microsoft/LightGBM/issues/2526#issuecomment-965137959)), so it'd definitely be great to have it as a general LightGBM parameter.

**Comment by jameslamb** [comment]
Ah right! Thank you for those references.

Alright I think we should make adding it to the `Config` on the C++ side part of the scope here @borchero . So that LightGBM won't raise a warning and so it'll survive the round-trip to/from model files.

**Comment by borchero** [comment]
Modifying the cpp config seemed to be straightforward enough :smile: I guess this is ready for review again (in case CI passes)

**Comment by borchero** [comment]
Tangentially related to this PR: do you think, we can do another release of LightGBM after this is merged @jameslamb? I'd really like to get my hands on #6314 and #6353 :smile: :eyes:

**Comment by jameslamb** [comment]
> *we can do another release of LightGBM after this is merged @jameslamb?*

Yeah I think it's a good time! It's been 3 months since the last one.

I'd love to get those fixes you mentioned out and #6019. I'll open a release PR where we can talk about what else we want to see get in before doing a release, and talk about whether it should be 4.3.1 or 4.4.0.

**Comment by jameslamb** [comment]
> I guess this is ready for review again

Could you please modify the tests I mentioned in https://github.com/microsoft/LightGBM/pull/6274#pullrequestreview-1945100771? If this parameter is expected to make it into model files, we should test that to be sure that isn't accidentally broken in the future.

**Comment by borchero** [comment]
I think this should do it. I also adjusted the C++ early stopping logic to make use of the min delta.

File: tests/python_package_test/test_engine.py
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1086]:
```suggestion
    gbm = lgb.train(params, lgb_train, feval=decreasing_metric, valid_sets=lgb_eval)
```

I *think* `valid_names` is optional, and that `lightgbm` will just substitute in names like `valid_0`, `valid_1`, etc.

Could you remove these, just to make this slightly more concise?

File: tests/python_package_test/test_engine.py
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1075]:
```suggestion
        "num_trees": num_trees,
        "num_leaves": 5,
```

Would you consider adding a small constraint on `num_leaves`, to make this even smaller and faster?

**Review by jameslamb — APPROVED** [review]
Looks great! I added 2 more very very minor comments on the tests... do what you want with those, and I don't need to review again. You can merge this after you've addressed them.

This is a really nice improvement, thanks for working on it!

Are you interested in doing a follow-up PR to add this same support to the R package? It'd be very similar.

https://github.com/microsoft/LightGBM/blob/14435485bd7e9c4d72ab43bd269c33fc4230212f/R-package/R/lgb.train.R#L258-L268

https://github.com/microsoft/LightGBM/blob/14435485bd7e9c4d72ab43bd269c33fc4230212f/R-package/R/callback.R#L187

If not, maybe @mayer79 would be interested (saw that he 👍🏻 'd some comments on this PR).

**Comment by borchero** [comment]
@jameslamb unfortunately, AppVeyor seems to be broken (also on `main`) which might be related to a faulty build on `conda-forge` (cf. https://anaconda.org/conda-forge/pyarrow/files: v15.0.2 build 5 was published right before our CI started failing)

**Comment by jameslamb** [comment]
@borchero I just applied these 2 small suggestions:

* https://github.com/microsoft/LightGBM/pull/6274#discussion_r1576910207
* https://github.com/microsoft/LightGBM/pull/6274#discussion_r1576910861

And I'll merge this PR once CI passes. If you disagree with either of them we can revert them, but I think they're non-controversial.

**Comment by borchero** [comment]
> @borchero I just applied these 2 small suggestions:

Sure! :smile: I would have taken care once CI was functional again 😬

**Comment by jameslamb** [comment]
> I would have taken care once CI 

yeah sorry, I was too eager 😬

**Comment by jameslamb** [comment]
I'm so glad to see this one finally merged! Thanks for all your hard work, @borchero

**Pull request merged**
